### ISSUE TITLE
VS2017 crash : GetTextViewLineRelativeToPoint

### ISSUE THREAD

**Issue by rkhoury** [VsVim/VsVim#2601]
**Describe the bug**
I had a random VS2017 crash that resulted in this Event Viewer log entry:
```
Application: devenv.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.ArgumentException
   at Microsoft.VisualStudio.Text.Editor.Implementation.TextViewLineCollection.ValidateBufferPosition(Microsoft.VisualStudio.Text.SnapshotPoint)
   at Microsoft.VisualStudio.Text.Editor.Implementation.TextViewLineCollection.Microsoft.VisualStudio.Text.Editor.IWpfTextViewLineCollection.GetTextViewLineContainingBufferPosition(Microsoft.VisualStudio.Text.SnapshotPoint)
   at Microsoft.VisualStudio.Text.Editor.Implementation.TextViewLineCollection.GetTextViewLineContainingBufferPosition(Microsoft.VisualStudio.Text.SnapshotPoint)
   at Vim.TextViewUtil.GetTextViewLineRelativeToPoint(Microsoft.VisualStudio.Text.Editor.ITextView, Int32, Microsoft.VisualStudio.Text.SnapshotPoint)
   at Vim.TextViewUtil.CheckScrollToPoint(Microsoft.VisualStudio.Text.Editor.ITextView, Microsoft.VisualStudio.Text.SnapshotPoint)
   at Vim.CommonOperations.EnsurePointVisible(Microsoft.VisualStudio.Text.SnapshotPoint)
   at Vim.CommonOperations.EnsureAtPointSync[[Vim.ViewFlags, Vim.Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]](Microsoft.VisualStudio.Text.SnapshotPoint, Vim.ViewFlags)
   at <StartupCode$Vim-Core>.$CommonOperations+EnsureAtCaret@2191[[Vim.ViewFlags, Vim.Core, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].Invoke(Microsoft.FSharp.Core.Unit)
   at Vim.UI.Wpf.VimHost+<>c__DisplayClass60_0.<DoActionWhenTextViewReady>g__doAction|0()
   at Vim.UI.Wpf.VimHost.DoActionWhenTextViewReady(Microsoft.FSharp.Core.FSharpFunc`2<Microsoft.FSharp.Core.Unit,Microsoft.FSharp.Core.Unit>, Microsoft.VisualStudio.Text.Editor.ITextView)
   at Vim.UI.Wpf.VimHost.Vim.IVimHost.DoActionWhenTextViewReady(Microsoft.FSharp.Core.FSharpFunc`2<Microsoft.FSharp.Core.Unit,Microsoft.FSharp.Core.Unit>, Microsoft.VisualStudio.Text.Editor.ITextView)
   at <StartupCode$Vim-Core>.$Vim+OnPositionChanged@657.Invoke(Microsoft.FSharp.Core.Unit)
   at <StartupCode$Vim-Core>.$CommonOperations+DoActionAsync@160.Invoke(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(System.Object)
   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
```

- Visual Studio version: 15.9.12
- VsVim version: 2.7.0.0
- Programming Language: C++
- Check(Type 'x') any that are installed:
    - [ ] ReSharper
    - [ ] Visual Assist

**Comment by ricksladkey** [comment]
Aw, darn it. I thought I caught all the ways that that could fail, but I missed one. Even worse, it's an exception that can bring down all of Visual Studio.

Thanks for the detailed bug report; I'll get this fixed right away.

**Comment by ricksladkey** [comment]
I made two changes to address this issue:

- Prevent exceptions when scheduling code from reaching the dispatcher
- Address this particular crash, which is a mismatch between the snapshot of the caret and the snapshot of the snapshot of the text view line collection

If you encounter this crash frequently, you can get (and test!) the fixed version before VsVim 2.8.0 is released by following these instructions:

- https://github.com/VsVim/VsVim/wiki/FAQ#how-can-i-install-a-build-with-the-latest-sources

**Comment by rkhoury** [comment]
Many thanks, @ricksladkey
